packages/mini-browser/src/browser/mini-browser-guard.ts

```diff
@@ -0,0 +1,114 @@
+
+import { MessageService } from '@theia/core';
+import { PreferenceService, PreferenceScope } from '@theia/core/lib/browser';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { MiniBrowserPreferences, IMiniBrowserPreferences } from './mini-browser-preferences';
+import { MiniBrowserConfiguration } from './mini-browser-configuration';
+
+/**
+ * Checks for known security issues with the mini-browser.
+ * Can be controlled through preferences.
+ */
+@injectable()
+export class MiniBrowserGuard {
+
+    @inject(MessageService)
+    protected messageService: MessageService;
+
+    @inject(PreferenceService)
+    protected preferenceService: PreferenceService;
+
+    @inject(MiniBrowserConfiguration)
+    protected miniBrowserConfiguration: MiniBrowserConfiguration;
+
+    @inject(MiniBrowserPreferences)
+    protected miniBrowserPreferences: MiniBrowserPreferences;
+
+    async onSetHostPattern(hostPattern: string): Promise<void> {
+        await this.miniBrowserPreferences.ready;
+        if (this.miniBrowserPreferences['mini-browser.warnIfUnsecure']) {
+            if (this.isHostPatternUnsecure(hostPattern)) {
+                this.messageService.warn(
+                    '`mini-browser` is currently configured to serve `file:` resources on the same origin as the application, this is known to be unsecure. ' +
+                    `Current pattern: \`${hostPattern}\``,
+                    { timeout: 5000 },
+                    /* actions: */ 'Ok', 'Don\'t show again',
+                ).then(action => {
+                    if (action === 'Don\'t show again') {
+                        this.setMiniBrowserPreference('mini-browser.warnIfUnsecure', false);
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Will throw if the location should not be opened, according to the current configurations.
+     */
+    async onFileLocationMap(location: string): Promise<void> {
+        await this.miniBrowserPreferences.ready;
+        if (this.isHostPatternUnsecure(this.miniBrowserConfiguration.hostPattern!)) {
+            if (this.miniBrowserPreferences['mini-browser.previewFile.preventUnsecure'] === 'alwaysPrevent') {
+                throw this.preventOpeningLocation(location);
+            }
+            if (this.miniBrowserPreferences['mini-browser.previewFile.preventUnsecure'] === 'ask') {
+                await this.askOpenFileUnsecurely(location);
+            }
+        }
+    }
+
+    protected isHostPatternUnsecure(hostPattern: string): boolean {
+        return hostPattern === '{{hostname}}';
+    }
+
+    protected async askOpenFileUnsecurely(location: string): Promise<void> {
+        const action = await this.messageService.warn(
+            'You are about to open a local file with the same origin as this application, this unsecure and the displayed document might access this application services. ' +
```

Comment:
"this unsecure" -> "this is potentially insecure"

"and the displayed document might access this application services."  -> remove? Or rephrase if you have an idea.